lp:1828532 scrub proxy relation data for all peers

kwmonroe opened the pull request:
See https://bugs.launchpad.net/layer-docker-registry/+bug/1828532

Way back in the 60s (revision 60s, that is), we added some invalid config on the website relation for both leader and follower units.  In the 80s (rev 87, #26), we corrected this with valid config, and in the early 100s (rev 101, #28), we sanitized the website relation data. Unfortunately, as the above bug shows, this only cleaned up the invalid config on leader units.

This PR does a similar cleanup for non-leader units.

axinojolais commented on reactive/docker_registry.py:
```diff
@@ -293,6 +293,19 @@ def update_reverseproxy_config():
+    website = endpoint_from_flag('website.available')
+    if not is_flag_set('charm.docker-registry.proxy-data.validated'):
+        website.set_remote(all_services=None, hostname=None, port=None)
+        set_flag('charm.docker-registry.proxy-data.validated')
```
Shouldn't you wrap this function with `@when_not('charm.docker-registry.proxy-data.validated')` ? I'm not sure I understand why the check is inside the function - which will make it call on every update-status hook as far as I can see.

kwmonroe replied:
@axinojolais yup.  I originally cargo'd this from the validation done on leader units. You're right that a `@when_not` guard is better.

This also made me think that when leadership changes, we'll need to make sure we send valid website relation data.  That is, if a leader becomes a follower, we'll need to scrub the `hostname` and `port` bits, as those are only valid from a leader-perspective.

I've added a commit to address these issues.

kwmonroe commented:
Released in rev 139.. Tests look good.

Start with docker-registry-64, known for injecting bad data on the relation:
```bash
$ juju deploy haproxy p64 -n 2
Located charm "cs:haproxy-48".
Deploying charm "cs:haproxy-48".

$ juju deploy cs:~containers/docker-registry-64 r64 -n 2
Located charm "cs:~containers/docker-registry-64".
Deploying charm "cs:~containers/docker-registry-64".

$ juju relate r64 p64

$ juju run --unit p64/0 'relation-ids reverseproxy'
reverseproxy:10

## bad data (all_services) from leader, r64/0
$ juju run --unit p64/0 'relation-get -r 10 - r64/0'
all_services: |2

  - service_name: r64
    service_host: 0.0.0.0
    service_port: 5000
    service_options:
     - mode http
     - balance leastconn
     - option httpchk GET / HTTP/1.0
    servers:
     - [r64-0, 172.31.1.251, 5000, 'check port 5000']
egress-subnets: 172.31.1.251/32
hostname: 172.31.1.251
ingress-address: 172.31.1.251
port: "5000"
private-address: 172.31.1.251

## bad data (all_services, hostname, and port) from follower, r64/1
$ juju run --unit p64/0 'relation-get -r 10 - r64/1'
all_services: |2

  - service_name: r64
    service_host: 0.0.0.0
    service_port: 5000
    service_options:
     - mode http
     - balance leastconn
     - option httpchk GET / HTTP/1.0
    servers:
     - [r64-1, 172.31.16.71, 5000, 'check port 5000']
egress-subnets: 172.31.16.71/32
hostname: 172.31.16.71
ingress-address: 172.31.16.71
port: "5000"
private-address: 172.31.16.71
```

Now upgrade to current stable to see leader cleaned up, yet follower still bad:
```bash
$ juju upgrade-charm r64 --channel stable
Added charm "cs:~containers/docker-registry-136" to the model.

## good data (s/all_services/services) from leader, r64/0
$ juju run --unit p64/0 'relation-get -r 10 - r64/0'
egress-subnets: 172.31.1.251/32
hostname: 172.31.1.251
ingress-address: 172.31.1.251
port: "5000"
private-address: 172.31.1.251
services: |2

  - service_name: r64
    service_host: 0.0.0.0
    service_port: 5000
    service_options:
     - mode http
     - balance leastconn
     - option httpchk GET / HTTP/1.0
    servers:
     - [r64-0, 172.31.1.251, 5000, check inter 2000 rise 2 fall 5 maxconn 4096]
     - [r64-1, 172.31.16.71, 5000, check inter 2000 rise 2 fall 5 maxconn 4096 backup]

## still bad data (all_services, hostname, and port) from follower, r64/1
$ juju run --unit p64/0 'relation-get -r 10 - r64/1'
all_services: |2

  - service_name: r64
    service_host: 0.0.0.0
    service_port: 5000
    service_options:
     - mode http
     - balance leastconn
     - option httpchk GET / HTTP/1.0
    servers:
     - [r64-1, 172.31.16.71, 5000, 'check port 5000']
egress-subnets: 172.31.16.71/32
hostname: 172.31.16.71
ingress-address: 172.31.16.71
port: "5000"
private-address: 172.31.16.71
```

Finally upgrade to 139, which contains this fix, to see good website data for all units:
```bash
$ juju upgrade-charm r64 --channel edge
Added charm "cs:~containers/docker-registry-139" to the model.

## still good data from the leader, r64/0
$ juju run --unit p64/0 'relation-get -r 10 - r64/0'
egress-subnets: 172.31.1.251/32
hostname: 172.31.1.251
ingress-address: 172.31.1.251
port: "5000"
private-address: 172.31.1.251
services: |2

  - service_name: r64
    service_host: 0.0.0.0
    service_port: 5000
    service_options:
     - mode http
     - balance leastconn
     - option httpchk GET / HTTP/1.0
    servers:
     - [r64-0, 172.31.1.251, 5000, check inter 2000 rise 2 fall 5 maxconn 4096]
     - [r64-1, 172.31.16.71, 5000, check inter 2000 rise 2 fall 5 maxconn 4096 backup]

## also good data (no service config) from the follower, r64/1
$ juju run --unit p64/0 'relation-get -r 10 - r64/1'
egress-subnets: 172.31.16.71/32
ingress-address: 172.31.16.71
private-address: 172.31.16.71
```

kwmonroe commented:
Per review, I updated flags to manage the website data for both lead and follower units better.  Released rev 140 and retested successfully.
